File: public/views/cards.js
suits = ['clubs', 'spades', 'hearts', 'diamonds']
ranks = ['2','3','4','5','6','7','8','9','10','jack','queen','king','ace']

var Deck = {}
Deck.cards = []
Deck.drawn_cards = []

isort_id = 0
for (var isuit in suits){
  for (var irank in ranks){
    Deck.cards.push([suits[isuit], ranks[irank], isort_id])
    isort_id += 1
  }
}

Deck.draw = function(){
  Deck.drawn_cards.push(Deck.cards.pop())
}

Deck.shuffle = function(){
  Deck.drawn_cards = shuffle(Deck.drawn_cards)
}

Deck.sort = function(){
  Deck.drawn_cards = Deck.drawn_cards.sort(function(obj1, obj2) {
    return obj1[2] - obj2[2];
});

}

Deck.draw()
Deck.draw()
Deck.draw()
Deck.shuffle()
Deck.sort()

for (var i in Deck.drawn_cards){
  console.log(Deck.drawn_cards[i])
}



function shuffle(array) {
  var currentIndex = array.length, temporaryValue, randomIndex;

  // While there remain elements to shuffle...
  while (0 !== currentIndex) {

    // Pick a remaining element...
    randomIndex = Math.floor(Math.random() * currentIndex);
    currentIndex -= 1;

    // And swap it with the current element.
    temporaryValue = array[currentIndex];
    array[currentIndex] = array[randomIndex];
    array[randomIndex] = temporaryValue;
  }

  return array;
}

File: public/views/cards.py
import random
 
class Deck(object):

  class Card(object):
    def __init__(self, suit, rank, sort_index):
      self.suit = suit 
      self.rank = rank
      self.sort_index = sort_index

    def __str__(self):
        return self.rank + ' of ' + self.suit

  def __init__(self):
    suits = ['clubs', 'spades', 'hearts', 'diamonds']
    ranks = ['2','3','4','5','6','7','8','9','10','jack','queen','king','ace']
    self.cards = []
    self.drawn_cards = []
    sort_index = 0
    for suit in suits:
      for rank in ranks:
        self.cards.append(self.Card(suit, rank, sort_index))
        sort_index += 1

  def draw(self):
    self.drawn_cards.append(self.cards.pop())
    return self.drawn_cards[-1]
  
  def shuffle(self, seed=None):
    if seed is not None: 
        random.Random(seed).shuffle(self.cards)
    else:
        shuffle(self.cards)

  def sort(self):
    self.cards.sort(key=lambda card: card.sort_index) 
